File: centrallix-os/sys/js/htdrv_scrollbar.js
// Copyright (C) 1998-2004 LightSys Technology Services, Inc.
//
// You may use these files and this library under the terms of the
// GNU Lesser General Public License, Version 2.1, contained in the
// included file "COPYING" or http://www.gnu.org/licenses/lgpl.txt.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.

function sb_init(param)
    {
    var tlayer=null;
    var l = param.layer;
    tlayer = htr_subel(l, param.tname);
    /*for(i=0;i<l.layers.length;i++)
	{
	var ml=l.layers[i];
	if(ml.name==param.tname) tlayer=ml;
	}*/
    var imgs = pg_images(l);
    for(i=0;i<imgs.length;i++)
	{
	var img=imgs[i];
	if(img.name=='d' || img.name=='u' || img.name=='b')
	    {
	    img.pane=l;
	    img.layer = img;
	    img.thum=tlayer;
	    img.kind='sb';
	    img.mainlayer=l;
	    }
	}
    imgs = pg_images(tlayer);
    imgs[0].kind='sb';
    imgs[0].layer = imgs[0];
    imgs[0].mainlayer=l;
    imgs[0].thum=tlayer;
    imgs[0].pane=l;
    tlayer.nofocus = true;
    htr_init_layer(tlayer, l, 'sb');
    htr_init_layer(l, l, 'sb');
    ifc_init_widget(l);
    l.thum = tlayer;
    //l.LSParent = param.parent;
    l.range = param.range;
    l.value = 0;
    l.is_horizontal = param.isHorizontal;
    l.controlsize = param.isHorizontal?(getClipWidth(l) - 18*3):(getClipHeight(l) - 18*3);
    l.sb_range_changed = sb_range_changed;
    htr_watch(l,'range','sb_range_changed');

    // Actions
    var ia = l.ifcProbeAdd(ifAction);
    ia.Add("MoveTo", sb_action_move_to);

    // Events
    var ie = l.ifcProbeAdd(ifEvent);
    ie.Add("MouseDown");
    ie.Add("MouseUp");
    ie.Add("MouseOver");
    ie.Add("MouseOut");
    ie.Add("MouseMove");

    l.SetThumb = sb_set_thumb;
    return l;
    }

function sb_action_move_to(aparam)
    {
    v = new Number(aparam.Value);
    v = v.valueOf();
    this.SetThumb(this.range, v);
    this.value = v;
    }

function sb_set_thumb(r,v)
    {
    if (this.is_horizontal)
	{
	if (r > 0) moveTo(this.thum, 18 + this.controlsize*v/(r), 0);
	}
    else
	{
	if (r > 0) moveTo(this.thum, 0, 18 + this.controlsize*v/(r));
	}
    }

function sb_range_changed(p,o,n)
    {
    if(cx__capabilities.Dom0IE)
	{
    	if(e.propertyName.substr(0,6) == "style.")
    	    {
	    n = e.srcElement.style[e.propertyName.substr(6)];
	    }
    	else
	    {
	    n = e.srcElement[e.propertyName];
	    }
	}
    // do the hokey pokey because 'n' might be a string (silly debug treeview...)
    n = new Number(n);
    n = n.valueOf();

    // Set thumb position and possibly the scrollbar's value
    if (n < 0) n = 0;
    //this.range = n;
    if (this.value > n) this.value = n;
    this.SetThumb(n,this.value);
    return n;
    }

function sb_do_mv()
    {
    var ti=sb_target_img;
    if (ti.kind=='sb' && sb_mv_incr != 0)
	{
	// Determine new offset
	var new_value = ti.mainlayer.value + sb_mv_incr;
	if (new_value > ti.mainlayer.range) new_value = Math.round(ti.mainlayer.range);
	if (new_value < 0) new_value = 0;

	// Set thumb position
	ti.mainlayer.SetThumb(ti.mainlayer.range, new_value);

	// Set offset
	ti.mainlayer.value = Math.round(new_value);
	}
    return true;
    }

function sb_tm_mv()
    {
    sb_do_mv();
    sb_mv_timeout=setTimeout(sb_tm_mv,50);
    return false;
    }

function sb_mousedown(e)
    {
    sb_target_img=e.target;
    if (sb_target_img != null && sb_target_img.kind=='sb' && (sb_target_img.name=='u' || sb_target_img.name=='d'))
        {
        if (sb_target_img.name=='u') sb_mv_incr=-10; else sb_mv_incr=+10;
        pg_set(sb_target_img,'src',htutil_subst_last(sb_target_img.src,"c.gif"));
        sb_do_mv();
        sb_mv_timeout = setTimeout(sb_tm_mv,300);
        return EVENT_HALT | EVENT_PREVENT_DEFAULT_ACTION;
        }
    else if (sb_target_img != null && sb_target_img.kind=='sb' && sb_target_img.name=='t')
        {
        sb_click_x = e.pageX;
        sb_click_y = e.pageY;
        sb_thum_x = getPageX(sb_target_img.thum);
        sb_thum_y = getPageY(sb_target_img.thum);
        return EVENT_HALT | EVENT_PREVENT_DEFAULT_ACTION;
        }
    else if (sb_target_img != null && sb_target_img.kind=='sb' && sb_target_img.name=='b')
        {
        sb_mv_incr=sb_target_img.mainlayer.controlsize + (18*3);
        if (!sb_target_img.mainlayer.is_horizontal && e.pageY < getPageY(sb_target_img.thum)+9) sb_mv_incr = -sb_mv_incr;
        if (sb_target_img.mainlayer.is_horizontal && e.pageX < getPageX(sb_target_img.thum)+9) sb_mv_incr = -sb_mv_incr;
        sb_do_mv();
        sb_mv_timeout = setTimeout(sb_tm_mv,300);
        return EVENT_HALT | EVENT_PREVENT_DEFAULT_ACTION;
        }
    else sb_target_img = null;
    if (e.kind == 'sb') cn_activate(e.mainlayer, 'MouseDown');
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function sb_mousemove(e)
    {
    var ti=sb_target_img;
    if (ti != null && ti.kind=='sb' && ti.name=='t')
        {
        if (ti.mainlayer.is_horizontal)
            {
            var new_x=sb_thum_x + (e.pageX-sb_click_x);
            if (new_x > getPageX(ti.pane)+18+ti.mainlayer.controlsize) new_x=getPageX(ti.pane)+18+ti.mainlayer.controlsize;
            if (new_x < getPageX(ti.pane)+18) new_x=getPageX(ti.pane)+18;
            setPageX(ti.thum,new_x);
            ti.mainlayer.value = Math.round((new_x - (getPageX(ti.pane)+18))*ti.mainlayer.range/ti.mainlayer.controlsize);
            }
        else
            {
            var new_y=sb_thum_y + (e.pageY-sb_click_y);
            if (new_y > getPageY(ti.pane)+18+ti.mainlayer.controlsize) new_y=getPageY(ti.pane)+18+ti.mainlayer.controlsize;
            if (new_y < getPageY(ti.pane)+18) new_y=getPageY(ti.pane)+18;
            setPageY(ti.thum,new_y);
            ti.mainlayer.value = Math.round((new_y - (getPageY(ti.pane)+18))*ti.mainlayer.range/ti.mainlayer.controlsize);
            }
        return EVENT_HALT | EVENT_PREVENT_DEFAULT_ACTION;
        }
    if (e.kind == 'sb') cn_activate(e.mainlayer, 'MouseMove');
    if (sb_cur_mainlayer && e.kind != 'sb')
        {
        cn_activate(sb_cur_mainlayer, 'MouseOut');
        sb_cur_mainlayer = null;
        }
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function sb_mouseup(e)
    {
    if (sb_mv_timeout != null)
        {
        clearTimeout(sb_mv_timeout);
        sb_mv_timeout = null;
        sb_mv_incr = 0;
        }
    if (sb_target_img != null)
        {
        if (sb_target_img.name != 'b')
            pg_set(sb_target_img,'src',htutil_subst_last(sb_target_img.src,"b.gif"));
        sb_target_img = null;
        }
    if (e.kind == 'sb') cn_activate(e.mainlayer, 'MouseUp');
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function sb_mouseover(e)
    {
    if (e.kind == 'sb')
        {
        if (!sb_cur_mainlayer)
            {
            cn_activate(e.mainlayer, 'MouseOver');
            sb_cur_mainlayer = e.mainlayer;
            }
        }
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

File: centrallix/htmlgen/htdrv_scrollbar.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include "ht_render.h"
#include "obj.h"
#include "cxlib/mtask.h"
#include "cxlib/xarray.h"
#include "cxlib/xhash.h"
#include "cxlib/mtsession.h"
#include "cxlib/strtcpy.h"

/************************************************************************/
/* Centrallix Application Server System 				*/
/* Centrallix Core       						*/
/* 									*/
/* Copyright (C) 1998-2003 LightSys Technology Services, Inc.		*/
/* 									*/
/* This program is free software; you can redistribute it and/or modify	*/
/* it under the terms of the GNU General Public License as published by	*/
/* the Free Software Foundation; either version 2 of the License, or	*/
/* (at your option) any later version.					*/
/* 									*/
/* This program is distributed in the hope that it will be useful,	*/
/* but WITHOUT ANY WARRANTY; without even the implied warranty of	*/
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the	*/
/* GNU General Public License for more details.				*/
/* 									*/
/* You should have received a copy of the GNU General Public License	*/
/* along with this program; if not, write to the Free Software		*/
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  		*/
/* 02111-1307  USA							*/
/*									*/
/* A copy of the GNU General Public License has been included in this	*/
/* distribution in the file "COPYING".					*/
/* 									*/
/* Module: 	htdrv_scrollbar.c					*/
/* Author:	Greg Beeley (GRB)					*/
/* Creation:	July 14, 2003    					*/
/* Description:	HTML Widget driver for a scrollbar - either horizontal	*/
/*		or vertical.						*/
/************************************************************************/


/** globals **/
static struct 
    {
    int		idcnt;
    }
    HTSB;


/*** htsbRender - generate the HTML code for the page.
 ***/
int
htsbRender(pHtSession s, pWgtrNode tree, int z)
    {
    char* ptr;
    char name[64];
    int x,y,w,h,r;
    int id, i, t;
    int visible = 1;
    char bcolor[64] = "";
    char bimage[64] = "";
    int is_horizontal = 0;
    pExpression code;

	if(!s->Capabilities.Dom0NS && !s->Capabilities.Dom0IE && !s->Capabilities.Dom1HTML)
	    {
	    mssError(1,"HTSB","Netscape 4.x, IE, or W3C DOM support required");
	    return -1;
	    }

    	/** Get an id for this. **/
	id = (HTSB.idcnt++);

	/** Which direction? **/
	if (wgtrGetPropertyValue(tree,"direction",DATA_T_STRING,POD(&ptr)) == 0)
	    {
	    if (!strcasecmp(ptr,"vertical")) is_horizontal = 0;
	    else if (!strcasecmp(ptr,"horizontal")) is_horizontal = 1;
	    }

    	/** Get x,y,w,h of this object **/
	if (wgtrGetPropertyValue(tree,"x",DATA_T_INTEGER,POD(&x)) != 0) 
	    {
	    mssError(1,"HTSB","Scrollbar widget must have an 'x' property");
	    return -1;
	    }
	if (wgtrGetPropertyValue(tree,"y",DATA_T_INTEGER,POD(&y)) != 0)
	    {
	    mssError(1,"HTSB","Scrollbar widget must have a 'y' property");
	    return -1;
	    }
	if (wgtrGetPropertyValue(tree,"width",DATA_T_INTEGER,POD(&w)) != 0)
	    {
	    if (is_horizontal)
		{
		mssError(1,"HTSB","Horizontal scrollbar widgets must have a 'width' property");
		return -1;
		}
	    else
		{
		w = 18;
		}
	    }
	if (is_horizontal && w <= 18*3)
	    {
	    mssError(1,"HTSB","Horizontal scrollbar width must be greater than %d", 18*3);
	    return -1;
	    }
	if (wgtrGetPropertyValue(tree,"height",DATA_T_INTEGER,POD(&h)) != 0)
	    {
	    if (!is_horizontal)
		{
		mssError(1,"HTSB","Vertical scrollbar widgets must have a 'height' property");
		return -1;
		}
	    else
		{
		h = 18;
		}
	    }
	if (!is_horizontal && h <= 18*3)
	    {
	    mssError(1,"HTSB","Vertical scrollbar height must be greater than %d", 18*3);
	    return -1;
	    }

	/** Get name **/
	if (wgtrGetPropertyValue(tree,"name",DATA_T_STRING,POD(&ptr)) != 0) return -1;
	strtcpy(name,ptr,sizeof(name));

	/** Range of scrollbar (static or dynamic property) **/
	if (is_horizontal)
	    r = w;
	else
	    r = h;
	if ((t = wgtrGetPropertyType(tree,"range")) == DATA_T_INTEGER)
	    {
	    wgtrGetPropertyValue(tree,"range",DATA_T_INTEGER,POD(&r));
	    }
	else if (t == DATA_T_CODE)
	    {
	    wgtrGetPropertyValue(tree,"range", DATA_T_CODE, POD(&code));
	    htrAddExpression(s, name, "range", code);
	    }

	/** Check background color **/
	if (wgtrGetPropertyValue(tree,"bgcolor",DATA_T_STRING,POD(&ptr)) == 0)
	    {
	    strtcpy(bcolor,ptr,sizeof(bcolor));
	    }
	if (wgtrGetPropertyValue(tree,"background",DATA_T_STRING,POD(&ptr)) == 0)
	    {
	    strtcpy(bimage,ptr,sizeof(bimage));
	    }

	/** Marked not visible? **/
	if (wgtrGetPropertyValue(tree,"visible",DATA_T_STRING,POD(&ptr)) == 0)
	    {
	    if (!strcmp(ptr,"false")) visible = 0;
	    }

	/** Ok, write the style header items. **/
	htrAddStylesheetItem_va(s,"\t#sb%POSpane { POSITION:absolute; VISIBILITY:%STR; LEFT:%INTpx; TOP:%INTpx; WIDTH:%POSpx; HEIGHT:%POSpx; clip:rect(0px,%POSpx,%POSpx,0px); Z-INDEX:%POS; }\n",id,visible?"inherit":"hidden",x,y,w,h,w,h, z);
	if (is_horizontal)
	    htrAddStylesheetItem_va(s,"\t#sb%POSthum { POSITION:absolute; VISIBILITY:inherit; LEFT:18px; TOP:0px; WIDTH:18px; Z-INDEX:%POS; }\n",id,z+1);
	else
	    htrAddStylesheetItem_va(s,"\t#sb%POSthum { POSITION:absolute; VISIBILITY:inherit; LEFT:0px; TOP:18px; WIDTH:18px; Z-INDEX:%POS; }\n",id,z+1);

	/** Write globals for internal use **/
	htrAddScriptGlobal(s, "sb_target_img", "null", 0);
	htrAddScriptGlobal(s, "sb_click_x","0",0);
	htrAddScriptGlobal(s, "sb_click_y","0",0);
	htrAddScriptGlobal(s, "sb_thum_x","0",0);
	htrAddScriptGlobal(s, "sb_thum_y","0",0);
	htrAddScriptGlobal(s, "sb_mv_timeout","null",0);
	htrAddScriptGlobal(s, "sb_mv_incr","0",0);
	htrAddScriptGlobal(s, "sb_cur_mainlayer","null",0);

	/** DOM Linkage **/
	htrAddWgtrObjLinkage_va(s, tree, "htr_subel(_parentctr, \"sb%POSpane\")",id);

	htrAddScriptInclude(s, "/sys/js/htdrv_scrollbar.js", 0);
	htrAddScriptInclude(s, "/sys/js/ht_utils_string.js", 0);

	/** Script initialization call. **/
	htrAddScriptInit_va(s,"    sb_init({layer:wgtrGetNodeRef(ns,\"%STR&SYM\"), tname:\"sb%POSthum\", isHorizontal:%INT, range:%INT});\n", name, id, is_horizontal, r);

	/** HTML body <DIV> elements for the layers. **/
	htrAddBodyItem_va(s,"<DIV ID=\"sb%POSpane\"><TABLE %[bgcolor=\"%STR&HTE\"%] %[background=\"%STR&HTE\"%] border=0 cellspacing=0 cellpadding=0 width=%POS>", id, *bcolor, bcolor, *bimage, bimage, w);
	if (is_horizontal)
	    {
	    htrAddBodyItem(s,   "<TR><TD align=right><IMG SRC=/sys/images/ico19b.gif width=18 height=18 NAME=u></TD><TD align=right>");
	    htrAddBodyItem_va(s,"<IMG SRC=/sys/images/trans_1.gif height=18 width=%POS name='b'>",w-36);
	    htrAddBodyItem(s,   "</TD><TD align=right><IMG SRC=/sys/images/ico18b.gif width=18 height=18 NAME=d></TD></TR></TABLE>\n");
	    }
	else
	    {
	    htrAddBodyItem(s,   "<TR><TD align=right><IMG SRC=/sys/images/ico13b.gif width=18 height=18 NAME=u></TD></TR><TR><TD align=right>");
	    htrAddBodyItem_va(s,"<IMG SRC=/sys/images/trans_1.gif height=%POS width=18 name='b'>",h-36);
	    htrAddBodyItem(s,   "</TD></TR><TR><TD align=right><IMG SRC=/sys/images/ico12b.gif width=18 height=18 NAME=d></TD></TR></TABLE>\n");
	    }
	htrAddBodyItem_va(s,"<DIV ID=\"sb%POSthum\"><IMG SRC=/sys/images/ico14b.gif NAME=t></DIV>",id);

	/** Add the event handling scripts **/

	htrAddEventHandlerFunction(s, "document","MOUSEDOWN","sb","sb_mousedown");
	htrAddEventHandlerFunction(s, "document","MOUSEMOVE","sb","sb_mousemove");
	htrAddEventHandlerFunction(s, "document","MOUSEUP","sb","sb_mouseup");
	htrAddEventHandlerFunction(s, "document", "MOUSEOVER", "sb","sb_mouseover");

	/** Check for more sub-widgets within the scrollbar (visual ones not allowed). **/
	for (i=0;i<xaCount(&(tree->Children));i++)
	    htrRenderWidget(s, xaGetItem(&(tree->Children), i), z+2);

	/** Finish off the last <DIV> **/
	htrAddBodyItem(s,"</DIV>\n");

    return 0;
    }


/*** htsbInitialize - register with the ht_render module.
 ***/
int
htsbInitialize()
    {
    pHtDriver drv;

    	/** Allocate the driver **/
	drv = htrAllocDriver();
	if (!drv) return -1;

	/** Fill in the structure. **/
	strcpy(drv->Name,"DHTML Scrollbar Widget Driver");
	strcpy(drv->WidgetName,"scrollbar");
	drv->Render = htsbRender;

	/** Events **/ 
	htrAddEvent(drv,"Click");
	htrAddEvent(drv,"MouseUp");
	htrAddEvent(drv,"MouseDown");
	htrAddEvent(drv,"MouseOver");
	htrAddEvent(drv,"MouseOut");
	htrAddEvent(drv,"MouseMove");

	htrAddAction(drv,"MoveTo");
	htrAddParam(drv,"MoveTo","Value",DATA_T_INTEGER);

	/** Register. **/
	htrRegisterDriver(drv);

	htrAddSupport(drv, "dhtml");
	HTSB.idcnt = 0;

    return 0;
    }
